drivers/include/periph/wdt.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2017 Technische Universität Berlin
+ *           (C) 2019 Inria
+ *           (C) 2019 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_periph_wdt WDT
+ * @ingroup     drivers_periph
+ * @brief       watchdog timer peripheral driver
+ *
+ * @{
+ *
+ * @file        wdt.h
+ * @brief       watchdog peripheral interface definitions
+ *
+ * @author      Thomas Geithner <thomas.geithner@dai-labor.de>
+ *              Francisco Molina <francois-xavier.molina@inria.fr>
+ *              Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#ifndef PERIPH_WDT_H
+#define PERIPH_WDT_H
+
+#include <stdint.h>
+#include "timex.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Signature for the watchdog early warning callback
+ *
+ * @param[in] arg           optional argument which is passed to the
+ *                          callback
+ */
+typedef void(*wdt_cb_t)(void *arg);
+
+/**
+ * @brief    Start watchdog timer
+ */
+void wdt_start(void);
+
+/**
+ * @brief    Stop watchdog timer
+ */
+void wdt_stop(void);
+
+/**
+ * @brief    Reset the watchdog timer counter, delay system reset
+ */
+void wdt_kick(void);
+
+/**
+ * @brief    Set up the wdt timer, only use max_time if normal operation
+ *           set min_time and max_time for windowed timer.
+ *
+ * @param[in] min_time       lower bound for windowed watchdog in us
+ * @param[in] max_time       upper bound for windowed watchdog in us, time before
+ *                           reset for normal watchdog
+ * @param[in] wdt_cb         wdt callback
```

Comment:
We could default to reboot, of no callback is given.

It might make sense to split this into to:

```
void wdt_setup_reboot(uint32_t min_time, uint32_t max_time);
void wdt_setup_callback(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg);
```

1. the latter can be guarded by a feature (e.g., periph_wdt_callback), should it not be available or desired
2. even if available but not used, the actual code will always have to reserve the 8 bytes of memory.

Reply:
At least on sam0 the Watchdog will always trigger a reboot, even when the early warning interrupt (callback) is enabled.

The callback is optional, when set it will be called a set time (e.g. one period) before the watchdog timer expires.

Reply:
In the case of stm32 the callback can feed/kick/reset de watchdog so a reboot is not the necessary outcome, it would make sense to me to split them having a separate function to setup the callback, which could be guarded.

Reply:
Well, we might have to make some concessions to a generic API here. The behaviour needs to be *the same* on all platforms implementing the interface. That's why I think it makes sense to split reboot and callback functionality.

How do other platforms handle callbacks/interrupts from the WDT?

(skimming datasheets)

STM32F4 seems to have a reset-only watchdog (IWDG) and a windowing one with reset + early warning interrupt.

nrf52 also has an early warning interrupt, but the datasheet says "Once the TIMEOUT event has been generated, the impending watchdog reset will always be effectuated.", suggesting that the watchdog cannot be reset from the interrupt.

Reply:
we could define the early warning interrupt in a way that all platforms support (e.g., it'll be triggered but cannot reset the watchdog), and emulate it using xtimer for platforms not supporting it...

Anyhow, I suggest we go with only ```wdt_setup_reset()``` for now. It'll be ok for 90% of the users, and we can extend the API later. What do you think?